### PR TITLE
SY-4491: Fix Arc On-Sources Stalling on Shared Index Channels

### PR DESCRIPTION
**Pull request by sy-nico**
# Issue Pull Request

## Linear Issue

[SY-4491](https://linear.app/synnax/issue/SY-4491)

## Description

The Arc `on` source paired each data sample with its index sample by buffer position. Because a shared index receives samples from every channel that references it, position N in the index buffer no longer lines up with position N in the data, so position pairing selected a mismatched index and the source returned without firing. Its high-water mark never advanced, so it stalled permanently. Anything gated on it (stage transitions, flow statements, log expressions) stopped re-evaluating even as values updated. Pair by alignment instead.

Adds a shared-index pairing suite covering the permanent-stall regression, sustained high-rate writes, late/missing index, high-water-mark ordering, index-less channels, and reset.

## Basic Readiness

- [X] I have performed a self-review of my code.
- [X] I have added relevant, automated tests to cover the changes.

### Greptile Summary

This PR fixes a permanent stall in the Arc `on` source when multiple channels share an index channel. The old code paired a data series with its index by buffer position (`indexData.Series[i]`), but a shared index accumulates extra series from every co-writing channel, so position `i` no longer corresponded to the correct timestamp — causing the alignment mismatch check to `return` without firing and permanently freezing the high-water mark.

* **Core fix** (`channels.go`): the new `indexSeriesFor` helper does a linear scan of the index buffer for an exact alignment match, decoupling the pairing from buffer position. The missing-index branch is also changed from `return` (permanent stall) to `continue` (skip to next data series), so one unresolved series no longer blocks all later ones.
* **Tests** (`channels_test.go`): 400 lines of new Ginkgo specs covering noise placement, exact timestamp pairing, N-channel shared-index concurrency, high-rate sustained writes, high-water-mark ordering, late/missing index, index-less channels, and reset.
* **Version**: patch bump 0.56.8 → 0.56.9.

### Confidence Score: 5/5

Safe to merge; the change is narrowly scoped to the index-pairing logic in one function and one new helper, and the new test suite directly exercises the regression plus all documented edge cases.

The alignment-based lookup in indexSeriesFor is a correct and minimal fix for the described stall. The continue replacing return on a missing index is an intentional complementary change. No issues were found that affect correctness.

No files require special attention.

### Important Files Changed

| Filename | Overview |
| -- | -- |
| arc/go/stl/channels/channels.go | Core fix: replaces position-based index pairing with alignment-based matching via new indexSeriesFor helper, correctly handling shared index buffers; also changes missing-index behavior from return (stall) to continue (try next series) |
| arc/go/stl/channels/channels_test.go | Adds 400-line test suite covering the permanent-stall regression, noise placement, timestamp precision, high-water-mark ordering, late/missing index, index-less channels, sustained high-rate writes, and reset |
| core/pkg/version/VERSION | Patch version bump 0.56.8 to 0.56.9, consistent with a bug-fix release |

### Flowchart

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[source.Next called] --> B[readSeries for channel key]
    B --> C{ok?}
    C -- No --> Z[return]
    C -- Yes --> D[Loop over data.Series]
    D --> E{ab.Lower < highWaterMark?}
    E -- Yes, stale --> D
    E -- No, fresh --> F{indexData type == UnknownT?}
    F -- Yes --> G[Synthesize timestamps via clock]
    F -- No --> H[indexSeriesFor: scan by alignment]
    H --> I{matching index found?}
    I -- No --> J[continue to next series]
    J --> D
    I -- Yes --> K[timeSeries = matched index series]
    G --> L[Output data + timeSeries]
    K --> L
    L --> M[highWaterMark = ab.Upper]
    M --> N[MarkChanged then return]
```

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A[source.Next called] --> B[readSeries for channel key]
    B --> C{ok?}
    C -- No --> Z[return]
    C -- Yes --> D[Loop over data.Series]
    D --> E{ab.Lower < highWaterMark?}
    E -- Yes, stale --> D
    E -- No, fresh --> F{indexData type == UnknownT?}
    F -- Yes --> G[Synthesize timestamps via clock]
    F -- No --> H[indexSeriesFor: scan by alignment]
    H --> I{matching index found?}
    I -- No --> J[continue to next series]
    J --> D
    I -- Yes --> K[timeSeries = matched index series]
    G --> L[Output data + timeSeries]
    K --> L
    L --> M[highWaterMark = ab.Upper]
    M --> N[MarkChanged then return]
```

Reviews (1): Last reviewed commit: ["SY-4491: Fix Arc On-Sources Stalling on ..."](<https://github.com/synnaxlabs/synnax/commit/3374c19b1273865b9ac3ffdf979e825c674a33de>) | [Re-trigger Greptile](<https://app.greptile.com/api/retrigger?id=44327820>)

### CODE REVIEW

File: arc/go/stl/channels/channels.go [hunk 1]
```diff
@@ -168,32 +168,44 @@ func (s *source) Next(ctx node.Context) {
...
+func indexSeriesFor(
+	indexData telem.MultiSeries,
+	alignment telem.Alignment,
+) (telem.Series, bool) {
+	for _, ser := range indexData.Series {
+		if ser.Alignment == alignment {
+			return ser, true
 		}
 	}
+	return telem.Series{}, false
```
Comment (pjdotson) [inline arc/go/stl/channels/channels.go]: alignments are ordered right? should probably do a binary search so it is faster.

File: arc/go/stl/channels/channels.go [hunk 2]
```diff
@@ -168,32 +168,44 @@ func (s *source) Next(ctx node.Context) {
...
+		}
+		var timeSeries telem.Series
+		if indexData.DataType() == telem.UnknownT {
+			timeSeries = telem.Arrange(
+				s.clock.Now(),
+				int(ser.Len()),
+				1*telem.NanosecondTS,
+			)
+			timeSeries.Alignment = ser.Alignment
+		} else if ts, found := indexSeriesFor(indexData, ser.Alignment); found {
```
Comment (pjdotson) [inline arc/go/stl/channels/channels.go]: maybe just imo but this is a little odd to have a helper function for. after this brace you can do the search, it's like 4 lines

**Review by pjdotson — CHANGES_REQUESTED** [review]
see comments

File: arc/go/stl/channels/channels.go (hunk 1 above)
Comment (sy-nico) [reply arc/go/stl/channels/channels.go]: The buffer isn't guaranteed sorted by alignment. It's built with `MultiSeries.Append`, which doesn't sort (only `NewMultiSeries` does), and nothing on the ingest/relay path sorts it either. A binary search would regress to the stall this fixes the first time delivery order isn't sorted. `ClearReads` keeps N small per cycle, and telem's own `MultiSeriesAtAlignment` is a linear scan for the same reason, so I kept it linear.

File: arc/go/stl/channels/channels.go (hunk 2 above)
Comment (sy-nico) [reply arc/go/stl/channels/channels.go]: Simply for readability. Updated the behavior and inlined.

**Review by pjdotson — APPROVED** [review]
